add-event.go

```diff
@@ -6,7 +6,7 @@ import (
 	"github.com/nbd-wtf/go-nostr"
 )
 
-func AddEvent(relay Relay, evt nostr.Event) (accepted bool, message string) {
+func AddEvent(relay Relay, evt nostr.Event) error {
```

Comment:
i would keep the `message string` in return values. per [nip-20](https://github.com/nostr-protocol/nips/blob/master/20.md). this use case comes to mind:

> Relays MUST return true when the event is a duplicate and has already been saved. The message SHOULD start with `duplicate:` in this case.

Reply:
i mean, instead of this complicated logic:

```go
if err.(*NostrError).Type == Duplicate {
  status = true
}
```

simply assume `err != nil` indicates `false` in command results, `true` otherwise. it's simpler to reason about when reading code.

Reply:
This is what it does: https://github.com/fiatjaf/relayer/pull/13/files#diff-56307bdb7ded67ebe516f562b90384407b21b2d34e8b5766b42602e1fb9705f0R152-R158

> Relays MUST return true when the event is a duplicate and has already been saved. The message SHOULD start with duplicate: in this case.

Exactly this was the inspiration to open the PR. :)

Reply:
using the `(err error, msg string)` return value, compare current patch with this:

```go
err, msg := AddEvent(relay, evt)
if err != nil {
  msg = err.Error()
}
ws.WriteJSON([]interface{}{"OK", evt.ID, err == nil, msg})
```

Reply:
(I've deleted my previous comment as you've posted a followup at the time when I've posted my reply.)

There's a problem due to the quoted condition:

> Relays MUST return true when the event is a duplicate and has already been saved. The message SHOULD start with duplicate: in this case.

And I wanted to avoid doing a direct string check for knowing if the `status` should be `true` or `false`, hence why I decide to surface the error instead and then decide at the upper level what the response should look like.

Maybe `AddEvent` should returning something like `func AddEvent(relay Relay, evt nostr.Event) (accepted bool, status bool, message string)` instead?

Reply:
> using the `(err error, msg string)` return value, compare current patch with this:
> 
> ```go
> err, msg := AddEvent(relay, evt)
> if err != nil {
>   msg = err.Error()
> }
> ws.WriteJSON([]interface{}{"OK", evt.ID, err == nil, msg})
> ```

Duplicate is an error so `err == nil` would result to `false` which shouldn't be the case as per NIP-20.

Reply:
> Duplicate is an error so err == nil would result to false which shouldn't be the case as per NIP-20.

if the rest of this patch remains as is, yes it would be invalid indeed. my suggestion was to update other parts to make the above snippet valid.

for example, this to me looks like a good place in add-event.go:

```go
saveErr := store.SaveEvent(&evt)
switch {
case errors.Is(saveErr, ErrDuplicate):
  return nil, "duplicate: ..." // duplicate is not an error in nip-20
case saveErr != nil:
  return err, err.Error()
}
// no errors; continue
```

again, `errors.Is(saveErr, ErrDuplicate)` is just an example. it wouldn't compile with this pull request as-is.

Reply:
This looks much cleaner indeed. I'll look into `Is()` and how those `SaveErr`, `DupErr` need to be structured.

Note that I'm only occasionally working with Go so I'm not aware of all of its concepts and best practices. Really appreciate your feedback on this.

Reply:
@x1ddos I've pushed https://github.com/fiatjaf/relayer/pull/13/commits/9a48626ec5958f75f8fd7360740b0f3f6e4cc5a7 - if we agre this is the right approach I'll squash and rebase the PR before it can be merged. Looking forward to hearing your feedback.

Reply:
looks definitely simpler. it's what i also had in mind. thanks!